asserts/model.go

```diff
@@ -25,16 +25,310 @@ import (
+	}
+
+	what := fmt.Sprintf("of snap %q", name)
+
+	snapID, err := checkStringMatchesWhat(snap, "id", what, validSnapID)
+	if err != nil {
+		return nil, err
+	}
+
+	typ, err := checkOptionalStringWhat(snap, "type", what)
+	if err != nil {
+		return nil, err
+	}
+	if typ == "" {
+		typ = "app"
+	}
+	if !strutil.ListContains(validSnapTypes, typ) {
+		return nil, fmt.Errorf("type of snap %q must be one of app|base|gadget|kernel|core", name)
+	}
+
+	modes, err := checkStringListInMap(snap, "modes", fmt.Sprintf("%q %s", "modes", what), validSnapMode)
+	if err != nil {
+		return nil, err
+	}
+
+	defaultChannel, err := checkOptionalStringWhat(snap, "default-channel", what)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: final name of this
+	track, err := checkOptionalStringWhat(snap, "track", what)
+	if err != nil {
+		return nil, err
+	}
+
+	if defaultChannel != "" && track != "" {
+		return nil, fmt.Errorf("snap %q cannot specify both default channel and locked track", name)
+	}
+	if track == "" && defaultChannel == "" {
+		defaultChannel = "stable"
+	}
+
+	if defaultChannel != "" {
+		_, err := channel.Parse(defaultChannel, "-")
+		if err != nil {
+			return nil, fmt.Errorf("invalid default channel for snap %q: %v", name, err)
+		}
+	} else {
+		trackCh, err := channel.ParseVerbatim(track, "-")
+		if err != nil || trackCh.Track == "" || trackCh.Risk != "" || trackCh.Branch != "" {
+			return nil, fmt.Errorf("invalid locked track for snap %q: %s", name, track)
+		}
+	}
+
+	presence, err := checkOptionalStringWhat(snap, "presence", what)
+	if err != nil {
+		return nil, err
+	}
+	if presence != "" && !strutil.ListContains(validSnapPresences, presence) {
+		return nil, fmt.Errorf("presence of snap %q must be one of required|optional", name)
+	}
+
+	return &ModelSnap{
+		Name:           name,
+		SnapID:         snapID,
+		SnapType:       typ,
+		Modes:          modes, // can be empty
+		DefaultChannel: defaultChannel,
+		Track:          track,
+		Presence:       presence, // can be empty
+	}, nil
+}
+
+// unextended case support
+
+func checkSnapWithTrack(headers map[string]interface{}, which string) (*ModelSnap, error) {
+	_, ok := headers[which]
+	if !ok {
+		return nil, nil
+	}
+	value, ok := headers[which].(string)
+	if !ok {
+		return nil, fmt.Errorf(`%q header must be a string`, which)
+	}
+	l := strings.SplitN(value, "=", 2)
+
+	name := l[0]
+	track := ""
+	if err := validateSnapName(l[0], which); err != nil {
+		return nil, err
+	}
+	if len(l) > 1 {
+		track = l[1]
+		if strings.Count(track, "/") != 0 {
+			return nil, fmt.Errorf(`%q channel selector must be a track name only`, which)
+		}
+		channelRisks := []string{"stable", "candidate", "beta", "edge"}
+		if strutil.ListContains(channelRisks, track) {
+			return nil, fmt.Errorf(`%q channel selector must be a track name`, which)
+		}
+	}
+
+	defaultChannel := ""
+	if track == "" {
+		defaultChannel = "stable"
+	}
+
+	return &ModelSnap{
```

Comment:
What about snap ID? The code in `checkModelSnap` checks it but this one does not. Perhaps this will be clearer as I read the rest of the diff below.

Reply:
this is the code for the old format, there is no snap id available there.

Reply:
Ah, this makes sense. +1